Lesson_21/homework_21.py

```diff
@@ -0,0 +1,135 @@
+        ' %(levelname)-7s - %(message)s',
+        '%Y-%m-%d %H:%M:%S',
+    )
+
+    # Create a handler for logging to the console
+    console_handler = logging.StreamHandler()
+    console_handler.setLevel(logging.DEBUG)
+    console_handler.setFormatter(log_formatter)
+
+    # Create a handler for logging to file
+    file_handler = logging.FileHandler(log_file, mode='a')
+    file_handler.setLevel(logging.DEBUG)
+    file_handler.setFormatter(log_formatter)
+
+    # Add the handler to the logger
+    logger.addHandler(console_handler)
+    logger.addHandler(file_handler)
+
+    return logger
+
+
+_log = configure_logging()
+
+
+def analyze_log(log_file, encoding='utf-8', key='Key TSTFEED0300|7E3E|0400'):
+    """Parse and analyze log file.
+
+    Args:
+        log_file (str): The path to the log file to be analyzed.
+        encoding (str, optional): The encoding used to read the log file.
+        key (str, optional): The key to filter the lines from the log file.
+    """
+    with open(log_file, 'rt', encoding=encoding) as l_fh:
+        log_file_lines = l_fh.readlines()
+
+    filtered_list = [line for line in log_file_lines if key in line]
+
+    parsed_time_list = [extract_timestamp(line) for line in filtered_list]
+
+    for i in range(len(parsed_time_list) - 1):
+        current_time = parsed_time_list[i]
+        next_time = parsed_time_list[i + 1]
+        log_difference(current_time, next_time)
+
+
+def extract_timestamp(line):
+    """Extract the timestamp and converts it to a `datetime` object.
+
+    Args:
+        line (str): A line from the log file that contains the timestamp.
+
+    Returns:
+        datetime: A `datetime` object representing the timestamp.
+    """
+    start_index = line.find('Timestamp ') + len('Timestamp ')
+    time_str = line[start_index:start_index + 8]  # Extract HH:MM:SS
+    return datetime.strptime(time_str, '%H:%M:%S')
+
+
+def log_difference(time_a, time_b):
+    """Compare two timestamps and logs an error or warning.
+
+    Args:
+        time_a (datetime): The first timestamp to compare.
+        time_b (datetime): The second timestamp to compare.
+
+    Returns:
+        float: The time difference in seconds between the two timestamps.
+    """
+    time_difference = (time_a - time_b).total_seconds()
+
+    time_a_str = time_a.strftime('%H:%M:%S')
+    time_b_str = time_b.strftime('%H:%M:%S')
+
+    if time_difference >= 33:
+        message = (f'Heartbeat between {time_a_str} and '
+                   f'{time_b_str} is {time_difference} seconds. '
+                   f'Greater than 33 seconds.')
+        return _log.error(message)
+    if 31 < time_difference < 33:
+        message = (f'Heartbeat between {time_a_str} and '
+                   f'{time_b_str} is {time_difference} seconds. '
+                   f'Greater than 31 seconds.')
+        return _log.warning(message)
+
+    return time_difference
+
+
+if __name__ == '__main__':
+    analyze_log('hblog.txt')
```

Comment:
Here you need to define not only file but certain key too. Yes you have value by default , but better to see that this option ( define key exists)

Reply:
Fix comments and rearrange to class in commit eaa7b1c132ccfbfe7fb21a3c9eea0cd25a09364b